### ISSUE TITLE
Unable to access Documents/Downloads/Desktop folders on MacOS when running brew services start code-server

### ISSUE THREAD

**Issue by uajqq** [cdr/code-server#2794]
- Web Browser: Safari
- Local OS: iPadOS/MacOS/iOS
- Remote OS: MacOS 11 Big Sur
- Remote Architecture: intel
- `code-server --version`: 3.9.2

`code-server` cannot view or edit any files in the Desktop, Documents, or Downloads folders, even after granting Terminal full disk privileges. There's no error, the folders just appear as empty. I can create files/folders through `code-server` in those locations, but any new file created there instantly disappears from `code-server`, while still being visible in Finder.

**Comment by bpmct** [comment]
> even after granting Terminal full disk privileges

For debugging purposes, what happens when you run code-server as root/sudo?

**Comment by uajqq** [comment]
I’ll try as soon as I’m able—I uninstalled `code-server` to try to see if reinstalling would help, and now I’m getting this error:

```
Installing from Homebrew.

+ brew install code-server
==> Downloading https://homebrew.bintray.com/bottles/icu4c-68.2.big_sur.bottle.t

curl: (35) error:1400410B:SSL routines:CONNECT_CR_SRVR_HELLO:wrong version number
Error: Failed to download resource "icu4c"
Download failed: https://homebrew.bintray.com/bottles/icu4c-68.2.big_sur.bottle.tar.gz
```

This looks more like a Homebrew error than a `code-server` error…

**Comment by uajqq** [comment]
Okay, so doing more experimentation:

Running `code-server` as root resolves the issue
Running `code-server` as myself also resolves the issue

Looks like this only happens when running `code-server` using `brew services start code-server`. I checked with `brew services list` and `code-server` is running with myself as the user.

**Comment by jsjoeio** [comment]
Glad to hear @bpmct's suggestion resolved the issue! Thanks for the extra info. We'll investigate further.

**Comment by uajqq** [comment]
Wait, the issue is not resolved--it persists when running code-server as a daemon via homebrew, which is how the official instructions say to do it. The only way to resolve this for me right is is to run it as a normal terminal command, which is not optimal.

**Comment by jsjoeio** [comment]
> Wait, the issue is not resolved

Sorry for the confusion on my part! You said `Running code-server as root resolves the issue` and `Running code-server as myself also resolves the issue`. When I say "resolved", I really meant it sounds like you're unblocked using these workarounds. 

We'll definitely look more into this when we have time!

**Comment by noahsbwilliams** [comment]
Newer versions of macOS require permission through a non-UNIX mechanism for access to the Desktop, Documents, Pictures, Downloads, and other folders.

You may have to give Node "full disk access" since it doesn't implement any of the macOS permission request stuff natively.

1. Find where Node is installed on your machine
```bash
➜  ~ which node
/usr/local/bin/node
```
2. Grant Node **Full Disk Access**

Open **System Preferences** > **Security & Privacy** > **Privacy** (horizontal) tab > **Full Disk Access** (vertical) tab > Click the 🔒 to unlock > Click **+** and select the Node binary you located.

**Comment by bpmct** [comment]
I created a PR that documents this in our FAQ. Should we leave this issue open or is this something beyond code-server's control that should be closed?

**Comment by wwk7225** [comment]
> Newer versions of macOS require permission through a non-UNIX mechanism for access to the Desktop, Documents, Pictures, Downloads, and other folders.
> 
> You may have to give Node "full disk access" since it doesn't implement any of the macOS permission request stuff natively.
> 
> 1. Find where Node is installed on your machine
> 
> ```shell
> ➜  ~ which node
> /usr/local/bin/node
> ```
> 
> 1. Grant Node **Full Disk Access**
> 
> Open **System Preferences** > **Security & Privacy** > **Privacy** (horizontal) tab > **Full Disk Access** (vertical) tab > Click the 🔒 to unlock > Click **+** and select the Node binary you located.

Well, it seems doesn't work for me...
After grant Full Disk Access permission for node binary, and restarted vs-code, the folders mentioned in this issue still cannot be accessed. Maybe some other binary file are also required to be granted Full Disk Access permision?

Another thing is that while executing `which node` in system terminal, it will echo `node not found`. But terminal in code-server can echo the correct path `/usr/local/opt/node@14/bin/node` (this path is not included in system's environment variable). Is it possible to make code-server share the same environment variable with system terminal (zsh)? The default `$PATH` value in code server doesn't consist MacTex's path `/Library/TeX/texbin`, which caused an error while building tex file.

To solve the permission issue, I edited the code-server's plist file created by `brew services` manually:
```
  <key>ProgramArguments</key>
  <array>
      <string>/usr/local/bin/fdautil</string>
      <string>exec</string>
      <string>/bin/zsh</string>
      <string>-c</string>
      <string>code-server</string>
  </array>
  <key>EnvironmentVariables</key>
  <dict>
      <key>PATH</key>
      <string>/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/TeX/texbin</string>
  </dict>
```
the 'fdautil' file is provided by an app called 'LaunchControl', which is also needed to granted with Full Disk Access permission previosly. I've also added the path for Tex in this plist file, and launch code-server by zsh. This one does work but not elegant at all, hopefully there could be a solution for original service management by homebrew.

**Comment by noahsbwilliams** [comment]
There is - hurrah!! The file you'd want to change is located [here](https://github.com/Homebrew/homebrew-core/blob/master/Formula/code-server.rb). I'm running short on time at the moment or I'd submit a PR myself.

**Comment by wwk7225** [comment]
> There is - hurrah!! The file you'd want to change is located [here](https://github.com/Homebrew/homebrew-core/blob/master/Formula/code-server.rb). I'm running short on time at the moment or I'd submit a PR myself.

Aha! It looks exactly the config file to generate plist file which should be edited (although I have no idea how to edit it permanently).
In my previous trying, I duplicated the plist file made by homebrew (then stop using homebrew to manage this service) and edit it directly with the content above. 
Well, I still have some trouble dealing with the environment varible. To solve the absence of path to TeX directory, I added the `EnvironmentVariables` part in plist file. But the `$PATH` value set in plist file seems have no effect if running code-server directly (the way which homebrew's original plist used), but launching code-server by zsh or bash will be fine. I have no idea how to add an environment variable which could work when running code-server directly (the path is already working in zsh and VScode, but not in code-server). Maybe another issue should be open...

**Comment by wwk7225** [comment]
> Newer versions of macOS require permission through a non-UNIX mechanism for access to the Desktop, Documents, Pictures, Downloads, and other folders.
> 
> You may have to give Node "full disk access" since it doesn't implement any of the macOS permission request stuff natively.
> 
> 1. Find where Node is installed on your machine
> 
> ```shell
> ➜  ~ which node
> /usr/local/bin/node
> ```
> 
> 1. Grant Node **Full Disk Access**
> 
> Open **System Preferences** > **Security & Privacy** > **Privacy** (horizontal) tab > **Full Disk Access** (vertical) tab > Click the 🔒 to unlock > Click **+** and select the Node binary you located.

I found an [answer](https://stackoverflow.com/questions/56203551/what-causes-an-eperm-error-when-traversing-my-home-directory) said:
> If you specifically need to access it in your Node application, you will need to enable full disk access for the application you're using to run this script (e.g, Terminal.app, iTerm, or an IDE) in System Preferences.

It seems that the application itself (code-server) but not node should be granted access to full disk? If code-server is launched in Terminal, the folder access permission is fine since most people have already granted terminal's access to Document/Desktop etc. But `brew services` use launchctl to start code-server directly, but code-server has not been granted permission.

I've also tried to grand permission to code-server (& node & Terminal), but it doesn't work. Still I think there is another binary file should be granted access to Document folder (or full disk).
